### DISCUSSION TITLE
API to change documents before executing plugins

### DISCUSSION THREAD

**Discussion by kazekyo** [dotansimha/graphql-code-generator#8593] [Ideas] [answered]
From: https://github.com/dotansimha/graphql-code-generator/issues/8296#issuecomment-1307011196

I want to use the new `client` preset, so I want to rewrite [my preset](https://github.com/kazekyo/nau) to a plugin.

`prepareDocuments` is what I need when I create a preset, isn't it? So it is different from my needs. 
Also, I don't think `prepareDocuments` has the features described below.

The first feature I need is to replace user-defined documents before executing plugins.
For example, my preset does the following replacements:
https://github.com/kazekyo/nau/blob/eec61904af1a826a20f5bf3fd63cdde02ac76aec/packages/graphql-codegen-preset/src/transforms/__tests__/passArgumentValueToFragment.test.ts#L6-L41
https://github.com/kazekyo/nau/blob/eec61904af1a826a20f5bf3fd63cdde02ac76aec/packages/graphql-codegen-preset/src/transforms/__tests__/addPaginationFields.test.ts#L6-L48

It is probably what your team calls "optimize" (optimizeOperation).
https://github.com/dotansimha/graphql-code-generator/blob/6470e6cc986274d450c405abeb9aa55c3fa6b6ab/packages/plugins/other/visitor-plugin-common/src/optimize-operations.ts#L5
Currently, the only plugin that can optimize documents is the relay-operation-optimizer. GraphQL Code Generator internally has a special implementation for relay-operation-optimizer.
https://github.com/dotansimha/graphql-code-generator/blob/d84afec0950b234cd716ea76436e3800a978328c/packages/plugins/typescript/operations/src/index.ts#L14-L18
One of the designs we could agree on might be to make the "optimize" a public API. 🤔 
```ts
const config: CodegenConfig = {
  schema: 'http://localhost:4000/graphql',
  documents: ['src/App.tsx'],
  generates: {
    './src/gql/': {
      preset: 'client',
      // ↓ Add this line ↓
      optimizers: ['@graphql-codegen/relay-operation-optimizer'], // or ['my-documents-optimizer', 'other-optimizer']
      plugins: ['my-plugin.ts'],
    },
  },
};
```

The second needed feature is an object that can pass data from the optimizer to the plugin. Because my plugin that generates some code needs unoptimized documents.

For example, suppose I wrote the following optimizer:
```js
module.exports = {
  optimizer(schema, documents, config) {
    const newDocuments = myChangeDocumentsFunction(documents);
    // ↓ Returns a context object
    return { documents: newDocuments, context: { myPlugin: { originalDocuments: documents } } };
  },
};
```
I want to refer to the context passed from the plugin.
```js
module.exports = {
  plugin(schema, documents, config, context) {
    return generateMyCode(schema, documents, context.myPlugin.originalDocuments);
  },
};
```

That is all!
I can create a demo implementation if it helps you understand.

**Comment by kazekyo** [comment, marked as answer]
@charlypoly What do you think? Or can you tell me who I should discuss it with?

**Reply by charlypoly** [reply]
@kazekyo, I think that you already have the support of many member of The Guild for your proposal 🎉 

Feel free to submit a first PR and put @saihaj and @n1ru4l as reviewers 👀

**Reply by kazekyo** [reply]
I understand. Thanks!

**Reply by kazekyo** [reply]
Resolved https://github.com/dotansimha/graphql-code-generator/pull/8723
